sei-tendermint/internal/p2p/transport.go

```diff
@@ -80,38 +92,113 @@ func HandshakeOrClose(
 			),
 		}, nil
 	})
 }
 
-func (c *Connection) Run(ctx context.Context) error {
-	return c.mconn.Run(ctx)
+func (c *Connection) sendRoutine(ctx context.Context, r *Router) error {
+	for {
+		start := time.Now().UTC()
+		m, err := c.sendQueue.Recv(ctx)
+		if err != nil {
+			return err
+		}
+		r.metrics.RouterPeerQueueRecv.Observe(time.Since(start).Seconds())
+		bz, err := proto.Marshal(m.Message)
+		if err != nil {
+			panic(fmt.Sprintf("proto.Marshal(): %v", err))
+		}
+		if m.ChannelID > math.MaxUint8 {
+			return fmt.Errorf("MConnection only supports 1-byte channel IDs (got %v)", m.ChannelID)
+		}
+		if err := c.mconn.Send(ctx, m.ChannelID, bz); err != nil {
+			return err
+		}
+		r.logger.Debug("sent message", "peer", c.peerInfo.NodeID, "message", m.Message)
+	}
 }
 
-// String displays connection information.
-func (c *Connection) String() string {
-	return c.RemoteEndpoint().String()
+// receivePeer receives inbound messages from a peer, deserializes them and
+// passes them on to the appropriate channel.
+func (c *Connection) recvRoutine(ctx context.Context, r *Router) error {
+	for {
+		chID, bz, err := c.mconn.Recv(ctx)
+		if err != nil {
+			return err
+		}
+		for chs := range r.channels.RLock() {
+			ch, ok := chs[chID]
+			if !ok {
+				// TODO(gprusak): verify if this is a misbehavior, and drop the peer if it is.
+				r.logger.Debug("dropping message for unknown channel", "peer", c.peerInfo.NodeID, "channel", chID)
+				continue
+			}
+
+			msg := proto.Clone(ch.desc.MessageType)
+			if err := proto.Unmarshal(bz, msg); err != nil {
+				return fmt.Errorf("message decoding failed, dropping message: [peer=%v] %w", c.peerInfo.NodeID, err)
+			}
+			if wrapper, ok := msg.(Wrapper); ok {
+				var err error
+				if msg, err = wrapper.Unwrap(); err != nil {
+					return fmt.Errorf("failed to unwrap message: %w", err)
+				}
+			}
+			// Priority is not used since all messages in this queue are from the same channel.
+			if _, ok := ch.recvQueue.Send(RecvMsg{From: c.peerInfo.NodeID, Message: msg}, proto.Size(msg), 0).Get(); ok {
+				r.metrics.QueueDroppedMsgs.With("ch_id", fmt.Sprint(chID), "direction", "in").Add(float64(1))
+			}
+			r.metrics.PeerReceiveBytesTotal.With(
+				"chID", fmt.Sprint(chID),
+				"peer_id", string(c.peerInfo.NodeID),
+				"message_type", r.lc.ValueToMetricLabel(msg)).Add(float64(proto.Size(msg)))
+			r.logger.Debug("received message", "peer", c.peerInfo.NodeID, "message", msg)
+		}
+	}
 }
 
-// SendMessage implements Connection.
-func (c *Connection) SendMessage(ctx context.Context, chID ChannelID, msg []byte) error {
-	if chID > math.MaxUint8 {
-		return fmt.Errorf("MConnection only supports 1-byte channel IDs (got %v)", chID)
+func (c *Connection) Run(ctx context.Context, r *Router) error {
+	peerInfo := c.PeerInfo()
+	peerID := peerInfo.NodeID
+	r.metrics.Peers.Add(1)
+	defer r.metrics.Peers.Add(-1)
+	for conns := range r.conns.Lock() {
+		if old, ok := conns[peerID]; ok {
+			old.Close()
```

Comment:
if we have an old connection here that we are closing the net connection for, how do we cleanup the scope.Run() with spawned send/recv routines and mconn? I'm not fully understanding the scenario in which we are creating a new connection and `Run`ning it while the router already has an old connection running? and if so, how do we ensure that the old one is being cleaned up properly?